src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -509,6 +509,8 @@ ModifyQuerySupported(Query *queryTree, bool multiShardQuery)
 	Node *onConflictWhere = NULL;
 
 	CmdType commandType = queryTree->commandType;
+	bool hasRecursivePlanning =
+		ContainsReadIntermediateResultFunction((Node *) queryTree);
```

Comment:
I think a query can contain an intermediate result AND another distributed table, but we don't do appropriate checks for the latter. Should this be `FindNodeCheck((Node *) queryTree, IsDistributedTableRTE)`?

Reply:
Should we error out for that case? And what type of a query are you thinking? I think we error out with having joins objection such as;

```sql
WITH select_data AS (
	SELECT * FROM modify_table
),
raw_data AS (
	UPDATE modify_table SET val = 0 WHERE 
		id IN (SELECT id FROM select_data) AND 
		val IN (SELECT counter FROM summary_table)
	RETURNING id, val
)
SELECT * FROM raw_data ORDER BY val;
```

```sql
ERROR:  cannot perform distributed planning for the given modification
DETAIL:  Joins are not supported in distributed modifications.
```

Reply:
Oh, that's good. I was thinking this is where we prevent joins. Do we need this check at all then? 

In other parts of the planner we treat subqueries like `(SELECT * FROM generate_series(...))` in the same way as `(SELECT * FROM read_intermediate_result(...))` and other subqueries that don't contain distributed tables. Could we do the same here? I.e. any subquery that doesn't contain distributed tables is acceptable.

Reply:
Do you mean something like this?
https://github.com/citusdata/citus/commit/cca756f81b9528849127530c3bea584699660f46#diff-4c4f0976727d06684376ecb22e8f5362

However, we still need to update `contain_volatile_functions ` calls;
https://github.com/citusdata/citus/commit/cca756f81b9528849127530c3bea584699660f46#diff-4c4f0976727d06684376ecb22e8f5362R721

Reply:
> Do you mean something like this?

I think we can actually just remove the whole `if (queryTree->hasSubLinks == true)` section, since we already check for joins later in that function. The error message changes a bit, but the error message is no longer accurate anyway.

> However, we still need to update contain_volatile_functions calls;

Yes, I think we need `contain_volatile_functions_citus` independently of this change.

Reply:
Example of what I mean:
https://github.com/citusdata/citus/commit/267c8038cb9e0b995e03e399ccfaa0bed933f4be

This allows us to run queries like:
```sql
WITH sel AS (
  SELECT DISTINCT y FROM test
), 
up AS (
  UPDATE test SET y = 1 WHERE x IN (SELECT * FROM sel) RETURNING *
)
SELECT count(*) FROM up;
````

as well as:
```sql
UPDATE test SET y = 1 WHERE x IN (SELECT 2*s FROM generate_series(1,10) s);
```

Reply:
> Example of what I mean:
> 267c803

Removing all subquery checks, we miss the upsert case. I will look into query tree after the lunch. Here is the example query:
https://github.com/citusdata/citus/pull/2006/commits/2673edca3adc94cd3215bfb572d268e07cdc1b9f

Reply:
Ah, makes sense, these error checks are a bit messy.

I guess you could just leave the hasSublinks checks and remove the `|| multiShardQuery` bit.

Reply:
Semi-related, the current error message is pretty weird:
```sql
INSERT INTO test VALUES (1,2) ON CONFLICT (x) DO UPDATE SET y = (SELECT count(*) FROM test);
ERROR:  subqueries are not supported in modifications across multiple shards
DETAIL:  Consider using an equality filter on partition column "x" to target a single shard.
```

Reply:
> I guess you could just leave the hasSublinks checks and remove the || multiShardQuery bit.

Note that, this will make `generate_series` error out.

Reply:
> Semi-related, the current error message is pretty weird:

Any suggestions which would work for all :)

Reply:
> Note that, this will make generate_series error out.

That's because of `hasRecursive`. You don't need `hasRecursive`.

Reply:
> Any suggestions which would work for all :)

I guess we could say: "subqueries in INSERT commands are not supported", not sure if this creates confusion w.r.t. INSERT...SELECT.

Reply:
> That's because of hasRecursive. You don't need hasRecursive.

I think I do, If I remove it, we will start to error out on subqueries which include CTEs. Such as;
https://github.com/citusdata/citus/commit/534a70c7985fce154dc067d39d349bec475c6fad#diff-acbd83aee5c8f2cf60953d2666f3963fR383

Reply:
> I guess we could say: "subqueries in INSERT commands are not supported", not sure if this creates confusion w.r.t. INSERT...SELECT.

But, we still error out on some UPDATE/DELETE queries.

Reply:
> I think I do, If I remove it, we will start to error out on subqueries which include CTEs. Such as;
534a70c#diff-acbd83aee5c8f2cf60953d2666f3963fR383

I don't follow. Where do you error out? 

See https://github.com/citusdata/citus/commit/7066cd720c5537b465fc2e3d762378b2904e1ab8

Reply:
I forgot to remove multishard part, now it seems good.